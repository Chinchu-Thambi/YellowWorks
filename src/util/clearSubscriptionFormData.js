/* globals localStorage */

import { storageAvailable } from '.';

const clearSubscriptionFormData = () => {
  if (storageAvailable()) {
    Object.keys(localStorage).forEach((key) => {
      if (
        key === 'storageOrderDetails'
        || key.startsWith('storageFormData')
        || key.startsWith('formData-') // still needed because onboarding uses it, yet.
      ) {
        localStorage.removeItem(key);
      }
    });
  }
};

export default clearSubscriptionFormData;
