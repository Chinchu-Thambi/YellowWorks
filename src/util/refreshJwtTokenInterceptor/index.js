import refreshJwtToken from './refreshJwtTokenInterceptor';

export default refreshJwtToken;
