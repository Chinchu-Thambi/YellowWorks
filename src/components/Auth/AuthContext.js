/* globals localStorage */

import React, {
  createContext, useReducer,
} from 'react';
import { navigate } from 'gatsby';
import { Hub } from '@aws-amplify/core';
import axios from 'axios';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import { NotificationManager } from 'react-notifications';

import useModule from '../../services/useModule';
import { storageAvailable, useLocalStorage } from '../../util';
import clearSubscriptionFormData from '../../util/clearSubscriptionFormData';
import { segmentIdentifyAndTrack } from '../../util/segment';

import getCustomerId from './services/getCustomerId';
import getStaffStatus from './services/getStaffStatus';

const initialState = {
  status: 'AUTHENTICATING',
  user: null,
  modal: {
    showAuthModal: false,
    // The target URL when this popup show up. Usually happens when accessing manage page directly
    params: null,
    target: null,
    dismissTarget: null,
  },
  customerId: null,
};

const reducer = (state, {
  type, user, error, modal,
  payload,
}) => {
  switch (type) {
    case 'AUTHENTICATED':
      return {
        ...state,
        status: 'AUTHENTICATED',
        user,
        modal: modal !== undefined ? modal : state.modal,
      };
    case 'UNAUTHENTICATED':
      return {
        ...state,
        status: 'UNAUTHENTICATED',
        error: error !== undefined ? error : state.error,
        modal: modal !== undefined ? modal : state.modal,
      };
    case 'AUTHENTICATING':
      return {
        ...state,
        status: 'AUTHENTICATING',
        modal: modal !== undefined ? modal : state.modal,
      };

    case 'LOGOUT':
      return {
        ...state,
        status: 'UNAUTHENTICATED',
        user: null,
      };

    case 'REGISTRATION_ERROR':
      return {
        ...state,
        error: error !== undefined ? error : state.error,
      };

    case 'SELECT_CUSTOMER':
      return {
        ...state,
        user: {
          ...state.user,
          selectedCustomer: payload.selectedCustomer,
        },
      };
    case 'TOGGLE_MODAL':
      return {
        ...state,
        modal: {
          ...state.modal,
          ...modal,
        },
      };
    default:
      return state;
  }
};

const AuthContext = createContext(null);

export default AuthContext;

let authState;

export const AuthProvider = ({ children }) => {
  const { Auth } = useModule('@aws-amplify/auth') || {};
  const [state, dispatch] = useReducer(reducer, initialState);
  const [selectedCustomer, setSelectedCustomer] = useLocalStorage('selectedCustomer', null);

  const jwtToken = R.path(['user', 'signInUserSession', 'idToken', 'jwtToken'])(state);
  const payload = R.path(['user', 'signInUserSession', 'idToken', 'payload'])(state);
  const expiryTime = R.path(['user', 'signInUserSession', 'idToken', 'payload', 'exp'])(state);
  const isStaff = getStaffStatus(state);
  const customerId = getCustomerId(state);

  // Gets current authentication state
  const getCurrentUser = async ({ bypassCache = false } = {}) => {
    try {
      dispatch({ type: 'AUTHENTICATING' });
      const currentUser = await Auth?.currentAuthenticatedUser({ bypassCache });

      dispatch({
        type: 'AUTHENTICATED', user: currentUser, modal: { showAuthModal: false },
      });
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(error);
      dispatch({ type: 'UNAUTHENTICATED', error });
    }
  };

  const refreshUser = React.useCallback(() => {
    if (!Auth) {
      return;
    }
    getCurrentUser({ bypassCache: true });
  }, [Auth]);

  const register = async (username, password, firstName, lastName) => {
    try {
      await Auth.signUp({
        username,
        password,
        attributes: {
          given_name: firstName,
          family_name: lastName,
        },
      });
      return true;
    } catch (error) {
      // TODO handle login error
      // eslint-disable-next-line no-console
      console.log(error);
      dispatch({ type: 'REGISTRATION_ERROR', error });
      return false;
    }
  };

  const confirmSignUp = async (username, code) => {
    try {
      const result = await Auth.confirmSignUp(username, code);
      return { success: true, result };
    } catch (error) {
      return { success: false, error };
    }
  };

  const resendCode = async (username) => {
    try {
      await Auth.resendSignUp(username);
      return { success: true };
    } catch (error) {
      // TODO handle login error
      // eslint-disable-next-line no-console
      console.log(error);
      return {
        success: false,
        error,
      };
    }
  };

  const login = async (username, password) => {
    try {
      const user = await Auth.signIn(username, password);
      dispatch({ type: 'AUTHENTICATED', user, modal: { showAuthModal: false } });
      const {
        attributes: {
          sub: id, email, given_name: givenName, family_name: familyName,
        },
      } = user;
      segmentIdentifyAndTrack({
        event: 'Login',
        id,
        payload: {
          email,
          firstmame: givenName,
          lastname: familyName,
        },
      });
      return { success: true, user };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = async () => {
    try {
      const { user } = state;
      if (user) {
        const {
          attributes: {
            sub: id, email, given_name: givenName, family_name: familyName,
          },
        } = user;
        segmentIdentifyAndTrack({
          event: 'Logout',
          id,
          payload: {
            email,
            firstname: givenName,
            lastname: familyName,
          },
        });
      }
      await Auth.signOut();

      clearSubscriptionFormData();

      navigate('/');
    } catch (error) {
      // TODO generic error handler
      // eslint-disable-next-line no-console
      console.log(error);
    }
    dispatch({ type: 'LOGOUT', modal: { showAuthModal: false } });
  };

  const forgotPassword = async (username) => {
    try {
      const result = await Auth.forgotPassword(username);
      return result;
    } catch (error) {
      // TODO generic error handler
      // eslint-disable-next-line no-console
      console.log(error);
      return { error };
    }
  };

  const forgotPasswordSubmit = async (username, code, newPassword) => {
    try {
      await Auth.forgotPasswordSubmit(username, code, newPassword);
      return true;
    } catch (error) {
      // TODO generic error handler
      // eslint-disable-next-line no-console
      console.log(error);
      return { error };
    }
  };

  const changePassword = async ({ previousPassword, proposedPassword }) => {
    const response = await Auth.changePassword(
      state.user,
      previousPassword,
      proposedPassword,
    );

    return response;
  };

  const loginGoogle = () => {
    if (storageAvailable()) {
      localStorage.setItem('yp_redirect', R.path(['modal', 'target'])(state));
      localStorage.setItem('yp_provider', 'Google');
      Auth.federatedSignIn({ provider: 'Google' });
    }
  };

  const loginFacebook = () => {
    if (storageAvailable()) {
      localStorage.setItem('yp_redirect', R.path(['modal', 'target'])(state));
      localStorage.setItem('yp_provider', 'Facebook');
      Auth.federatedSignIn({ provider: 'Facebook' });
    }
  };

  const loginYellow = () => {
    if (storageAvailable()) {
      localStorage.setItem('yp_redirect', R.path(['modal', 'target'])(state));
      localStorage.setItem('yp_provider', 'Yellow');
      Auth.federatedSignIn({ provider: 'Yellow' });
    }
  };

  const showHideModal = ({
    show, modalType, target, params, dismissTarget, dismissable,
  }) => {
    dispatch({
      type: 'TOGGLE_MODAL',
      modal: {
        modalType, showAuthModal: show, target, params, dismissTarget, dismissable,
      },
    });
  };

  const displayLoginModal = ({ target, dismissTarget, dismissable } = {}) => {
    showHideModal({
      show: true,
      modalType: 'login',
      target,
      dismissTarget,
      dismissable,
    });
  };

  const getCustomerListByEmail = async (email) => {
    const { data: { data, errors } } = await axios.post(
      process.env.GATSBY_APPSYNC_URL,
      {
        query: `query {
          getCustomerListByEmail(email: "${email}"){
            name
            companyId
          }
        }`,
      },
      {
        headers: {
          Authorization: jwtToken,
        },
      },
    );

    if (errors) {
      // eslint-disable-next-line no-console
      console.error('ERRORS', errors);
      throw errors;
    }

    return data;
  };

  const selectCustomer = React.useCallback(async (customerObject) => {
    if (!customerObject || !isStaff) return;

    const { contactId, customerId: selectedCustomerId, customerName } = customerObject;

    setSelectedCustomer({
      contactId,
      customerId: selectedCustomerId,
      customerName,
    });
  }, [isStaff, setSelectedCustomer]);

  // Configure Auth
  React.useEffect(() => {
    if (!Auth) {
      return;
    }

    Auth.configure({
      region: `${process.env.GATSBY_AWS_REGION}`,
      userPoolId: `${process.env.GATSBY_COGNITO_USER_POOL_ID}`,
      userPoolWebClientId: `${process.env.GATSBY_COGNITO_POOL_CLIENT_ID}`,
      oauth: {
        domain: `${process.env.GATSBY_OAUTH_DOMAIN}`,
        scope: ['email', 'profile', 'openid', 'aws.cognito.signin.user.admin'],
        redirectSignIn: `${process.env.GATSBY_OAUTH_REDIRECT_SIGN_IN}`,
        redirectSignOut: `${process.env.GATSBY_OAUTH_REDIRECT_SIGN_OUT}`,
        responseType: 'code',
      },
    });
  }, [Auth]);

  // Gets initial authentication state
  // Second argument of empty array means this effect will only run once
  React.useEffect(() => {
    if (!Auth) {
      return;
    }

    getCurrentUser();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [Auth]);

  // Redirects user post social login
  React.useEffect(() => {
    if (storageAvailable() && jwtToken) {
      const redirectUrl = localStorage.getItem('yp_redirect');
      localStorage.removeItem('yp_redirect');
      localStorage.removeItem('yp_provider');
      // After redirecting from Social Login, log to Segment
      const {
        sub: id,
        email,
        given_name: givenName,
        family_name: familyName,
      } = state.user?.attributes || {};
      segmentIdentifyAndTrack({
        event: 'Login',
        id,
        payload: {
          email,
          firstname: givenName,
          lastname: familyName,
        },
      });
      if (redirectUrl && redirectUrl !== 'undefined') {
        navigate(redirectUrl);
      }
    }
  }, [jwtToken, state]);

  // set listener for post social auth
  React.useEffect(() => {
    if (
      !Auth
    ) { return () => {}; }

    let isPending = false;
    const hubListener = async (data) => {
      const { payload: { event } } = data;

      if (event === 'signIn' && !isPending) {
        isPending = true;

        if (!state.user) {
          const currentUser = await Auth.currentAuthenticatedUser();
          dispatch({ type: 'AUTHENTICATED', user: currentUser });
        }
        isPending = false;
      }
    };

    Hub.listen('auth', hubListener);

    return () => {
      Hub.remove('auth', hubListener);
    };
  }, [Auth, state.user]);

  // refreshes user session when it expires
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    const now = new Date();
    const expiryDate = new Date(expiryTime * 1000);
    const delay = expiryDate - now;

    const timeout = setTimeout(async () => {
      refreshUser();
    }, delay);

    return () => { clearTimeout(timeout); };
  }, [expiryTime, refreshUser]);

  // replaces jwtToken when a customerId is selected
  React.useEffect(() => {
    if (!Auth) {
      return;
    }

    const getUserToken = async () => {
      if (
        isStaff
        && selectedCustomer?.customerId
        && selectedCustomer?.customerId !== customerId
      ) {
        const { username } = state.user;

        const challengedUser = await Auth.signIn(username);

        const user = await Auth.sendCustomChallengeAnswer(
          challengedUser,
          jwtToken,
          {
            contactId: selectedCustomer.contactId,
            customerId: selectedCustomer.customerId,
          },
        );

        if (getCustomerId({ user }) === selectedCustomer.customerId) {
          dispatch({ type: 'AUTHENTICATED', user });
        } else {
          setSelectedCustomer({ customerId });
          NotificationManager.error(
            `Customer ID ${selectedCustomer.customerId} couldn't be found in HubSpot. `
            + 'Please make sure you\'ve got the right one.',
          );
        }
      }
    };

    getUserToken();
  }, [Auth, customerId, isStaff, jwtToken, selectedCustomer, setSelectedCustomer, state.user]);

  authState = {
    // Auth Attributes
    state,
    jwtToken,
    payload,
    customerId,
    isStaff,

    // AWS Cognito Auth
    getCurrentUser,
    refreshUser,
    register,
    confirmSignUp,
    resendCode,
    login,
    logout,
    forgotPassword,
    forgotPasswordSubmit,
    changePassword,
    loginGoogle,
    loginFacebook,
    loginYellow,

    // Modal
    showHideModal,
    displayLoginModal,

    // Staff Tools
    getCustomerListByEmail,
    selectCustomer,
    selectedCustomer,
  };

  return (
    <AuthContext.Provider value={authState}>
      {children}
    </AuthContext.Provider>
  );
};

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
