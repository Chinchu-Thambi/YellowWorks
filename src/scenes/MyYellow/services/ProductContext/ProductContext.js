import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import * as R from 'ramda';
import { NotificationManager } from 'react-notifications';

import AuthContext from '../../../../components/Auth/AuthContext';
import { useLocalStorage } from '../../../../util';

const ProductContext = React.createContext();
export default ProductContext;

export const ProductProvider = ({ children }) => {
  const { jwtToken, customerId } = React.useContext(AuthContext) || {};
  const [subscriptionId, setSubscriptionId] = React.useState(null);
  const [storageFormData, setStorageFormData] = useLocalStorage('storageFormData', null);
  const [storageOrderDetails, setStorageOrderDetails] = useLocalStorage('storageOrderDetails', null);
  const [storageFormDataExpiry, setStorageFormDataExpiry] = useLocalStorage('storageFormDataExpiry', {});
  const [isLoading, setIsLoading] = React.useState(false);
  const [storageErrors, setStorageErrors] = useLocalStorage('storageFormDataErrors', null);
  const formData = storageFormData?.[subscriptionId] || null;
  const orderDetails = storageOrderDetails?.[subscriptionId] || null;
  const expiryTime = storageFormDataExpiry?.[subscriptionId] || null;
  const errors = storageErrors?.[subscriptionId] || null;

  const expireFormData = React.useCallback(() => {
    delete storageFormDataExpiry[subscriptionId];
    setStorageFormDataExpiry({
      ...storageFormDataExpiry,
    });
  }, [setStorageFormDataExpiry, storageFormDataExpiry, subscriptionId]);

  const setExpiryDate = React.useCallback((delay) => {
    const now = new Date();
    const futureDate = new Date(now.getTime() + delay).toISOString();
    setStorageFormDataExpiry({
      ...storageFormDataExpiry,
      [subscriptionId]: futureDate,
    });
  }, [setStorageFormDataExpiry, storageFormDataExpiry, subscriptionId]);

  const saveData = async ({ path, data }) => {
    const updatedFormData = R.assocPath(
      path,
      data,
    )(formData);

    if (!R.equals(updatedFormData, formData)) {
      const query = `
        mutation submitSchema(
          $subscriptionId: ID!
          $customerId: ID!
          $input: AWSJSON!
          $publish: Boolean!
        ) {
          updateConfiguration(
            subscriptionId: $subscriptionId
            customerId: $customerId
            input: $input
            publish: $publish
          )
        }
      `;

      const variables = {
        subscriptionId,
        customerId,
        input: JSON.stringify(updatedFormData),
        publish: true,
      };

      const response = await axios.post(
        process.env.GATSBY_PRODUCT_PLATFORM_ENDPOINT,
        { query, variables },
        { headers: { Authorization: jwtToken } },
      );

      const submitProductBrief = R.compose(
        R.tryCatch(JSON.parse, () => ({})),
        R.path(['data', 'data', 'updateConfiguration']),
      )(response);

      if (submitProductBrief.error) {
        console.error('There was a problem with the user input.', {
          key: `subscription_${subscriptionId}`,
          variables,
          orderDetails,
        });
        NotificationManager.error('Input could not be saved.', '', 5000);
        return { success: false };
      }

      setStorageFormData({
        ...storageFormData,
        [subscriptionId]: updatedFormData,
      });

      const in3Seconds = 3 * 1000;
      setExpiryDate(in3Seconds);

      NotificationManager.success('Saved successfully.', '', 5000);

      return { success: true };
    }

    return { success: false };
  };

  // load formData
  React.useEffect(() => {
    // bare requirements for formData loading
    if (
      !jwtToken
      || !subscriptionId
      || isLoading
    ) return;

    // controls for data based loading necessity
    if (
      expiryTime && formData
    ) return;

    const tomorrow = (24 * 60 * 60 * 1000);
    setExpiryDate(tomorrow);
    setStorageErrors({
      ...storageErrors,
      [subscriptionId]: null,
    });

    const getSubscriptionContent = async () => {
      const query = `
        query getProductContextData($id: ID!, $customerId: ID!) {
          subscription(
            id: $id
            customerId: $customerId
          ) {
              id
              orderDetails {
                updated
                activated
                product {
                  sku
                  id
                  name
                  configuration
                  pricingPlan
                }
                productOptions {
                  name
                  id
                  sku
                  configuration
                  pricingPlan
                }
                providerData {
                  slug
                }
                status
                userSchema
              }
              productDetails {
                userConfiguration
                status
              }
              upcomingOrderDetails {
                updated
                activated
                product {
                  sku
                  id
                  name
                  configuration
                  pricingPlan
                }
                productOptions {
                  name
                  id
                  sku
                  configuration
                  pricingPlan
                }
                status
              }
            }
          }
      `;
      const variables = {
        id: subscriptionId,
        customerId,
      };

      if (expiryTime) {
        setStorageFormData({
          ...storageFormData,
          [subscriptionId]: undefined,
        });
        setStorageOrderDetails({
          ...storageOrderDetails,
          [subscriptionId]: undefined,
        });
      }
      setIsLoading(true);

      const response = await axios.post(
        process.env.GATSBY_PRODUCT_PLATFORM_ENDPOINT,
        { query, variables },
        { headers: { Authorization: jwtToken } },
      );

      if (response.data?.errors) {
        setStorageErrors({ [subscriptionId]: response.data.errors });
      }

      const orderDetailsData = R.path(['data', 'data', 'subscription', 'orderDetails'])(response);
      const userSchema = R.tryCatch(JSON.parse, () => {
        setStorageErrors({
          [subscriptionId]: [
            ...response.data.errors,
            {
              errorType: 'Schema parsing',
              message: 'Schema could not be parsed.',
            },
          ],
        });
        return null;
      })(orderDetailsData?.userSchema);
      orderDetailsData.userSchema = userSchema;

      const data = R.compose(
        R.tryCatch(JSON.parse, () => {
          setStorageErrors({
            [subscriptionId]: [
              ...response.data.errors,
              {
                errorType: 'Data parsing',
                message: 'Data could not be parsed.',
              },
            ],
          });
          return null;
        }),
        R.path(['data', 'data', 'subscription', 'productDetails', 'userConfiguration']),
      )(response);

      setStorageFormData({
        ...storageFormData,
        [subscriptionId]: data,
      });
      setStorageOrderDetails({
        ...storageOrderDetails,
        [subscriptionId]: orderDetailsData,
      });
      setIsLoading(false);
    };
    getSubscriptionContent();
  }, [customerId, expiryTime, formData, isLoading, jwtToken, orderDetails, setStorageErrors, setExpiryDate,
    setStorageFormData, setStorageOrderDetails, storageFormData, storageOrderDetails, subscriptionId, storageErrors]);

  // setup reload for existing expiration time
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    const now = new Date();
    const expiryDate = new Date(expiryTime);
    const delay = expiryDate - now;

    const timeout = setTimeout(() => {
      expireFormData();
    }, delay);

    return () => { clearTimeout(timeout); };
  }, [expireFormData, expiryTime]);

  // sets expiration time for pending listing
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    if (orderDetails?.status === 'PENDING') {
      const delay = 5000;
      const timeout = setTimeout(() => {
        expireFormData();
      }, delay);

      return () => { clearTimeout(timeout); };
    }

    return () => {};
  }, [expireFormData, expiryTime, orderDetails]);

  // sets expiration time for when an error happens
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    if (errors) {
      const in20Seconds = 20 * 1000;
      const timeout = setTimeout(() => {
        expireFormData();
      }, in20Seconds);

      return () => { clearTimeout(timeout); };
    }

    return () => {};
  }, [errors, expireFormData, expiryTime]);


  const productState = {
    // status
    subscriptionId,
    errors,

    // data
    formData,
    orderDetails,

    // action
    saveData,
    setSubscriptionId,
    expireFormData,
  };

  return (
    <ProductContext.Provider
      value={productState}
    >{children}
    </ProductContext.Provider>
  );
};

ProductProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
