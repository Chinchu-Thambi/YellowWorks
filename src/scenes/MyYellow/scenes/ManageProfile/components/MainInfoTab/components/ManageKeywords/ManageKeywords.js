import React from 'react';
import * as R from 'ramda';
import Loadable from 'react-loadable';
import { Box } from 'rebass';

import tagIcon from '../../../../../../../../assets/icons/light/tag.svg';

import Button from '../../../../../../../../components/Button';
import Modal from '../../../../../../../../components/Modal';
import Spinner from '../../../../../../../../components/Spinner';
import Tag from '../../../../../../../../components/Tag';

import ProductContext from '../../../../../../services/ProductContext';

import {
  ListItem, OptionItem, LabelItem, ActionButton, OptionDetails,
} from '../../../../../../components/ListGroup';

import { ButtonContainer, EditWrapper } from '../../../modals/Styled';

import getKeywordsFromCategories from './services/getKeywordsFromCategories';

const ManageKeywordsForm = Loadable({
  loader: () => import('./components/ManageKeywordsForm'),
  loading: Spinner,
});

const path = ['categories'];

const ManageKeywords = () => {
  const productState = React.useContext(ProductContext) || {};
  const categories = R.pathOr([], path)(productState.formData);
  const keywordsFromCategories = getKeywordsFromCategories(categories);

  const [modalVisibility, setModalVisibility] = React.useState(false);
  const [localData, setLocalData] = React.useState(categories);
  const [pendingSave, setPendingSave] = React.useState(false);

  const toggleVisibility = () => {
    setModalVisibility(!modalVisibility);
  };

  const handleSave = async () => {
    setPendingSave(true);
    const { success } = await productState.saveData({
      path,
      data: localData,
    });
    setPendingSave(false);

    if (success) {
      toggleVisibility();
    }
  };

  const handleChange = React.useCallback((formData) => {
    if (!R.equals(localData, formData)) {
      setLocalData(formData);
    }
  }, [localData]);

  // updates localState if categories change.
  React.useEffect(() => {
    if (!R.equals(localData, categories)) {
      setLocalData(categories);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [categories]);

  return (
    <>
      <ListItem>
        <OptionItem>
          <LabelItem icon={tagIcon}>Keywords</LabelItem>
          <div>
            <ActionButton
              variant="link"
              size="sm"
              onClick={toggleVisibility}
            >
              Edit
            </ActionButton>
          </div>
        </OptionItem>
        <OptionDetails>
          {keywordsFromCategories?.map(({ keyword }) => (
            <Tag key={keyword} readonly>{keyword}</Tag>
          ))}
        </OptionDetails>
      </ListItem>

      {modalVisibility && (
        <Modal
          title="Keywords"
          onDismiss={toggleVisibility}
          isVisible
        >
          <EditWrapper>
            <ManageKeywordsForm
              value={localData}
              onChange={handleChange}
            />

            <ButtonContainer>
              {pendingSave && (
                <Box alignSelf="center" mr={2}>
                  <Spinner size={16} mr={2} />
                </Box>
              )}
              <Button onClick={handleSave} disabled={pendingSave} size="sm">save</Button>
            </ButtonContainer>
          </EditWrapper>
        </Modal>
      )}
    </>
  );
};

export default ManageKeywords;
