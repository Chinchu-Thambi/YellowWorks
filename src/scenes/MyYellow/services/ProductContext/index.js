import ProductContext, { ProductProvider } from './ProductContext';

export { ProductProvider };
export default ProductContext;
