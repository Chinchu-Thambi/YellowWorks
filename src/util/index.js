/* global window, DOMException, document */

import React from 'react';
import showdown from 'showdown';

import { importableModules } from '../services/useModule';

export {
  default as theme,
  mediaBreakpointUp,
} from './theme';
export {
  default as errorText,
  infos as infoText,
  getErrorString,
} from './texts';

const isSingleLine = (string) => (string !== undefined ? !string.includes('\n') : '');

const mdConverter = new showdown.Converter();

mdConverter.setOption('tables', true);

export const parseMd = (md) => {
  if (!md) {
    return '';
  }

  let mdString = md;

  if (typeof md === 'object') {
    [mdString] = Object.values(md);
  }

  let parsedString = mdConverter.makeHtml(mdString);

  if (isSingleLine(mdString)) {
    // remove outer p tag from single line md
    parsedString = parsedString.replace(/<\/?p>/g, '');
  }

  // eslint-disable-next-line react/no-danger
  return <span dangerouslySetInnerHTML={{ __html: parsedString }} />;
};

export const getProportionalHeight = ({
  initialWidth, initialHeight, targetWidth,
}) => (initialHeight * targetWidth) / initialWidth;

export const storageAvailable = () => {
  let storage;

  if (typeof window !== 'undefined') {
    storage = window.localStorage;
  }

  try {
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    if (typeof DOMException === 'undefined') {
      return false;
    }

    return e instanceof DOMException && (
      // everything except Firefox
      e.code === 22
      // Firefox
      || e.code === 1014
      // test name field too, because code might not be present
      // everything except Firefox
      || e.name === 'QuotaExceededError'
      // Firefox
      || e.name === 'NS_ERROR_DOM_QUOTA_REACHED')
      // acknowledge QuotaExceededError only if there's something already stored
      && storage && storage.length !== 0;
  }
};

export const formatPrice = (price = 0) => (typeof price === 'number' ? `$${price.toFixed(2)}` : `$${price}`);

export const formatRecurringPrice = (price, recurring, withGst, annual) => {
  const priceFormatted = formatPrice(price);
  if (priceFormatted === '$0.00') {
    return 'Free';
  }

  if (recurring) {
    if (withGst) {
      return `${priceFormatted} / month + GST`;
    }
    if (annual) {
      return `${priceFormatted} / year`;
    }
    return `${priceFormatted} / month`;
  }
  if (withGst) {
    return `${priceFormatted} + GST`;
  }
  return priceFormatted;
};

// ex input date object
// ex out: "13/08/2019"
export const reformatDate = (dateObj) => (
  dateObj ? dateObj.toLocaleDateString('en-NZ') : ''
);

export const useLocalStorage = (key, initialValue) => {
  // from https://usehooks.com/useLocalStorage/
  // State to store our value
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = React.useState(() => {
    try {
      // Get from local storage by key
      if (storageAvailable()) {
        const item = window.localStorage.getItem(key);
        // Parse stored json or if none return initialValue
        return item ? JSON.parse(item) : initialValue;
      }
      return initialValue;
    } catch (error) {
      // If error also return initialValue
      console.log(error);
      return initialValue;
    }
  });

  // Return a wrapped version of useState's setter function that ...
  // ... persists the new value to localStorage.
  const setValue = (value) => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      if (storageAvailable()) {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
    } catch (error) {
      // A more advanced implementation would handle the error case
      console.log(error);
    }
  };

  return [storedValue, setValue];
};

// TODO: remove this and user version inside of auth/services
export const isStaff = async () => {
  try {
    const { Auth } = await importableModules['@aws-amplify/auth']();
    const user = await Auth.currentSession();
    return user.idToken.payload.email.endsWith('@yellow.co.nz');
  } catch (err) {
    // console.log('isStaff', err);
    return false;
  }
};


export const isTesting = () => document.location.search.includes('utm_isTesting=yes');

export const getNextWeekday = () => {
  const dt = new Date();
  dt.setDate(dt.getDate() + 1);

  if (dt.getDay() === 6) { // if Saturday
    dt.setDate(dt.getDate() + 2);
  }

  if (dt.getDay() === 0) { // if Sunday
    dt.setDate(dt.getDate() + 1);
  }

  return dt;
};

export const getNextMonth = ({
  baseNextWeekday = false,
} = {}) => {
  let dt = new Date();

  if (baseNextWeekday) {
    dt = getNextWeekday();
  }

  dt.setFullYear(
    dt.getFullYear(),
    dt.getMonth() + 1,
    1,
  );

  return dt;
};

export const getEndOfMonth = ({
  baseNextWeekday = false,
} = {}) => {
  let dt = new Date();

  if (baseNextWeekday) {
    dt = getNextWeekday();
  }

  dt.setFullYear(
    dt.getFullYear(),
    dt.getMonth() + 1,
    0,
  );

  return dt;
};

export const capitalize = (string = '') => `${string.charAt(0).toUpperCase()}${string.substr(1).toLowerCase()}`;

export const shuffle = (array) => {
  let counter = array.length;
  const shuffled = [...array];

  // While there are elements in the array
  while (counter > 0) {
    // Pick a random index
    const index = Math.floor(Math.random() * counter);

    // Decrease counter by 1
    counter -= 1;

    // And swap the last element with it
    const temp = shuffled[counter];
    shuffled[counter] = shuffled[index];
    shuffled[index] = temp;
  }

  return shuffled;
};
