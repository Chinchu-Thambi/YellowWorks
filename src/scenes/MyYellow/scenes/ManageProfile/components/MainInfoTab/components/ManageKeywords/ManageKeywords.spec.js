/* globals describe, it, expect, jest, beforeAll, afterAll */

import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, act, fireEvent } from '@testing-library/react';
import * as R from 'ramda';
import axios from 'axios';

import AuthContext from '../../../../../../../../components/Auth/AuthContext';
import { theme } from '../../../../../../../../util';

import ProductContext from '../../../../../../services/ProductContext';

import ManageKeywords from './ManageKeywords';

describe('ManageKeywords', () => {
  beforeAll(() => {
    jest.setTimeout(30000);
  });

  afterAll(() => {
    jest.setTimeout(5000);
  });

  const keyword = 'mockkeyword';

  const productState = {
    formData: {
      categories: [{
        id: 1,
        keywords: [{
          keyword,
        }],
      }],
    },
    saveData: jest.fn().mockResolvedValue({ success: true }),
  };

  it('renders without throwing', () => {
    expect(async () => {
      render(
        <ThemeProvider theme={theme}>
          <ManageKeywords />
        </ThemeProvider>,
      );
    }).not.toThrow();
  });

  it('handles keywords', async () => {
    const wrapper = await render(
      <ThemeProvider theme={theme}>
        <AuthContext.Provider value={{ jwtToken: 'jwtToken' }}>
          <ProductContext.Provider value={productState}>
            <ManageKeywords />
          </ProductContext.Provider>
        </AuthContext.Provider>
      </ThemeProvider>,
    );

    await wrapper.findByText('Keywords');
    const editButton = await wrapper.findByText('Edit');
    await wrapper.findByText(keyword);

    axios.post.mockResolvedValueOnce(
      {
        data: {
          data: {
            category1: JSON.stringify(R.assocPath(
              ['properties', 'keywords', 'items', 'enum'],
              [{
                keyword: 'BRANZ',
                property: 'Associations',
              }],
            )({})),
          },
        },
      },
    );

    await act(async () => {
      fireEvent.click(editButton);
    });

    let removableKeywords = await wrapper.findAllByLabelText('Remove');
    expect(removableKeywords).toHaveLength(1);

    const selectInput = await wrapper.findByLabelText('Select the Products & Services associated with your business:');

    await act(async () => {
      fireEvent.change(selectInput, { target: { value: 'BR' } });
    });

    await act(async () => {
      fireEvent.click(await wrapper.findByText('Associations: BRANZ'));
    });

    removableKeywords = await wrapper.findAllByLabelText('Remove');
    expect(removableKeywords).toHaveLength(2);

    await act(async () => {
      fireEvent.click(removableKeywords[1]);
    });

    removableKeywords = await wrapper.findAllByLabelText('Remove');
    expect(removableKeywords).toHaveLength(1);

    await act(async () => {
      fireEvent.click(await wrapper.findByText('save'));
    });

    expect(productState.saveData.mock.calls[0]).toMatchSnapshot();
  });
});
