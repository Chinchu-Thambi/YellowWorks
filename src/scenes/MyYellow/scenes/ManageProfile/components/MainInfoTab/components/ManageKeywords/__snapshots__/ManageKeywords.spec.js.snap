// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageKeywords handles keywords 1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "id": 1,
        "keywords": Array [
          Object {
            "keyword": "BRANZ",
            "property": "Associations",
          },
        ],
      },
    ],
    "path": Array [
      "categories",
    ],
  },
]
`;
